vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold_authorization.cc

```diff
@@ -89,152 +83,63 @@ void UpholdAuthorization::Authorize(
   }
 
-  // After a login, we want to attempt to relink the user's payment ID to their
-  // Uphold wallet address. Clear the flag that will cause relinking to be
-  // skipped.
-  ledger_->state()->SetAnonTransferChecked(false);
-
-  ledger_->uphold()->SetWallet(wallet_ptr->Clone());
-
-  auto user_callback = std::bind(&UpholdAuthorization::OnGetUser,
-      this,
-      _1,
-      _2,
-      callback);
-  ledger_->uphold()->GetUser(user_callback);
-}
-
-void UpholdAuthorization::OnGetUser(
-    const type::Result result,
-    const User& user,
-    ledger::ExternalWalletAuthorizationCallback callback) {
-  auto wallet_ptr = GetWallet(ledger_);
-  base::flat_map<std::string, std::string> args;
-
-  if (user.bat_not_allowed || !wallet_ptr) {
-    BLOG(0, "BAT not allowed");
-    callback(type::Result::BAT_NOT_ALLOWED, args);
-    return;
+  if (uphold_wallet->status != type::WalletStatus::NOT_CONNECTED &&
+      uphold_wallet->status != type::WalletStatus::DISCONNECTED_VERIFIED) {
+    return callback(type::Result::LEDGER_OK, {});
   }
 
-  if (user.status == UserStatus::OK) {
-    wallet_ptr->status = user.verified
-        ? type::WalletStatus::VERIFIED
-        : type::WalletStatus::CONNECTED;
-    ledger_->uphold()->SetWallet(wallet_ptr->Clone());
-
-    if (wallet_ptr->address.empty()) {
-      auto new_callback = std::bind(&UpholdAuthorization::OnCardCreate,
-          this,
-          _1,
-          _2,
-          callback);
-      ledger_->uphold()->CreateCard(new_callback);
-      return;
-    }
+  DCHECK(uphold_wallet->token.empty());
+  DCHECK(uphold_wallet->address.empty());
 
-    if (!user.verified) {
-      args["redirect_url"] = GetSecondStepVerify();
-    }
-  } else {
-    wallet_ptr->status = type::WalletStatus::PENDING;
-    ledger_->uphold()->SetWallet(wallet_ptr->Clone());
-    args["redirect_url"] = GetSecondStepVerify();
+  if (result != type::Result::LEDGER_OK) {
+    BLOG(0, "Couldn't exchange code for the access token!");
+    return callback(result, {});
   }
 
-  callback(type::Result::LEDGER_OK, args);
-}
-
-void UpholdAuthorization::OnCardCreate(
-    const type::Result result,
-    const std::string& address,
-    ledger::ExternalWalletAuthorizationCallback callback) {
-  if (result == type::Result::LEDGER_ERROR) {
-    BLOG(0, "Card creation");
-    callback(type::Result::LEDGER_ERROR, {});
-    return;
+  if (token.empty()) {
+    BLOG(0, "Access token is empty!");
+    return callback(type::Result::LEDGER_ERROR, {});
   }
 
-  auto wallet_ptr = GetWallet(ledger_);
-  wallet_ptr->address = address;
-  ledger_->uphold()->SetWallet(wallet_ptr->Clone());
-
-  if (!address.empty()) {
-    ledger_->database()->SaveEventLog(
-        log::kWalletConnected,
-        static_cast<std::string>(constant::kWalletUphold) + "/" +
-            address.substr(0, 5));
+  const auto from_status = uphold_wallet->status;
+  uphold_wallet->status = type::WalletStatus::PENDING;
+  const auto to_status = uphold_wallet->status;
```

Comment:
By the time we do the logging, `uphold_wallet` is already in a moved-from state, so this won't work:
```
ledger_->database()->SaveEventLog(
    log::kWalletStatusChange, (std::ostringstream{} << from_status).str() +
                                  " ==> " +
                                  (std::ostringstream{} << uphold_wallet->status).str());
```
This could work though:
```
ledger_->database()->SaveEventLog(
    log::kWalletStatusChange, (std::ostringstream{} << from_status).str() +
                                  " ==> " +
                                  (std::ostringstream{} << type::WalletStatus::PENDING).str());
```
but I wanted the logged status to reflect the value in `uphold_wallet->status`, so that if we change the implementation of `UpholdAuthorization::OnAuthorize()` at some point in the future, they don't get out of sync.